### PR TITLE
core: add flatten-rbd-pvc command

### LINKED ISSUE #222: Add a command to flatten RBD PVC
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

Add a command to flatten RBD PVC.

**What is use case behind this feature:**

- Increase the durability of RBD PVC
- Prevent performance degradation of RBD PVC from other PVCs  
- Make it possible to mirror cloned RBD PVC

This feature will reduce the operation cost to flatten RBD PVC by hand like this.

1. FInd the name of PV correspoinding to the target PVC.
2. Get the `volumeHandle` file of the PV
3. Get the name of the target RBD image from the `volumeHandle` field
4. Flatten the target RBD image.

**Related issues and discussions:**
[RBD Async: Failed to mirrored Cloned PVC created from another PVC (PVC-PVC Clone)](https://github.com/ceph/ceph-csi/issues/2426) in ceph-csi
[RBD Async: Failed to mirrored Cloned PVC created from snapshot (PVC from snapshot)](https://github.com/ceph/ceph-csi/issues/2427) in ceph-csi
[How to achieve mirrirong cloned PVCs](https://github.com/ceph/ceph-csi/discussions/4360) in ceph-csi

### PR DESCRIPTION
**Pull request by satoru-takeuchi**
**Description of your changes:**

- Flatten an RBD image corrensponding to the target PVC
- Remove the corresponding temporary RBD image

**Which issue is resolved by this Pull Request:**
Resolves #222

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] CI tests has been updated, if necessary.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.

### CODE REVIEW

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:2]: 2024

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:49]: Can we add extra checks like below

* Ensure PVC is bound and not under deletion

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:56]: This code panics if the PV is not a CSI PVC, you also need to if imageName and pool properties are set in the `VolumeAttributes` are not

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:60]: dont we need to add `return` here?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:63]: Consider the `radosnamepsace` as well or we can open an issue and track it

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:73]: This can be a heavy operation, How about if we do below

* Get the rbd image `info`
* Move the image to trash
* Add a task for the image Id which we got in `info` output` (in case if the CLI restarts due to panic or user action) we might end up leaving an image in the trash, i believe ceph will have some internal periodic garbage collection of cleanup of stale images from trash.

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:24]: Add a new line for 3rd party and internal imports?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:41]: How about adding `Long` description of what it does internally for `temp` image as well?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:78]: add a check to ensure that its a cloned rbd image or return an error if its not cloned? This check can be done in two places

* Check PVC is having DataSource as PVC
* With rbd `info` status?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:57]: This only works for the PVC with DataSouce PVC what about the PVC created from the snapshot?

**Review by Madhu-1 — COMMENTED** [review]
@satoru-takeuchi i left some early suggestion (even for draft PR) please let me know if it makes sense :)

**Comment by satoru-takeuchi** [comment]
@Madhu-1 Thank you for your valuable comments! I'll reflect them later.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: I'll use `rbd info` because it can verify the target image is certainly a cloned image.

**Comment by subhamkrai** [comment]
@satoru-takeuchi any update on this?

**Comment by satoru-takeuchi** [comment]
@subhamkrai Oops, sorry. I overlooked your message. I'll update this PR to be ready to review in this week.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: I'll open a new issue to support `radosnamespace`.

**Comment by satoru-takeuchi** [comment]
Tests has not been finished yet. I'll udate this PR the beginning of the next week.

**Comment by satoru-takeuchi** [comment]
I lost working-on branch by mistake. Please wait for a more bit..

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: I changed my mind. Checking `DataSource` is enough.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: Done.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: Changed the code not to delete the temporary clone image iff the target RBD PVC is cloned from PVC.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: Opened an issue #276 .

**Comment by subhamkrai** [comment]
> @Madhu-1 I reflected all your comments. I have two questions about testing.
> 
> * Q1. unit test: Do unit tests necessary for the new command? There isn't any test code under cmd/commands rather than root_test.go.

@satoru-takeuchi not all the commands required unit test, generally when we string passing in helper function we are adding unit tests but overall not really.

> * Q2. integration test: Is it OK to just add a test to flatten a cloned-from-PVC image for both "default-namespace" and "custom-namespace"? Or should I add more tests (e.g. cloned-from-snapshot, try to flatten in-use PVC, and so on) to improve coverage? The current other code seems to test only happy pathes.

@Madhu-1 could you take q2? Thanks

**Comment by Madhu-1** [comment]
> * Q2. integration test: Is it OK to just add a test to flatten a cloned-from-PVC image for both "default-namespace" and "custom-namespace"? Or should I add more tests (e.g. cloned-from-snapshot, try to flatten in-use PVC, and so on) to improve coverage? The current other code seems to test only happy pathes.

Sorry for delay as i was on PTO, Lets not block this PR for it but good to have tests to cover all cases, we can open a new issue to add more tests to ensure it works for all the cases

**Comment by satoru-takeuchi** [comment]
@subhamkrai @Madhu-1 Thank you for your comments. I'm trying on adding an integration test. I'll make this PR ready to review after that.

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:71]: Add few more validations here as below
* Check if its CSI PVC
* Check if its RBD PVC

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:72]: use short variable declaration `shouldDeleteTempImage:=false` ?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:90]: "Can we log what is exactly missing? This will help in debugging." like `PV %s doesn't contains imageName in VolumeAttributes`

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:94]: same as above.

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:104]: `if len(status.Watchers) > 0 {` looks to be more readable, WDYT?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:57]: point to devel branch?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:113]: flatten takes time depending on the data size, should be make use of `ceph rbd task add flatten`?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:136]: there is a corner case where CLI killed in between moving to trash and add trash remove, if someone issues the info command it fails and it will be left in trash, we can enhance it later but a pointer to be noted here.

File: tests/github-action-helper.sh
Comment (Madhu-1) [inline tests/github-action-helper.sh:271]: couple of things
* declare version as a variable that needs update in only one place 
* use the latest 7.0.2 release
* download the folder to temp folder instead of the root from where the tool is running.

File: tests/github-action-helper.sh
Comment (Madhu-1) [inline tests/github-action-helper.sh:280]: i think you are referring to pvc-pvc clone here not pvc-restore right?

where are we creating the volumesnapshotclass and volumesnapshot, if its planned in different PR can we make the changes related to it in same PR instead of here?

**Review by Madhu-1 — COMMENTED** [review]
some nits, @satoru-takeuchi you have covered most of the cases 👏🏻

**Comment by satoru-takeuchi** [comment]
@Madhu-1 Thank you for your comments. I'm back from PTO. I'll update this PR tomorrow.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go:104]: I agree with you. Fixed.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go:57]: Fixed to point main branch.

File: tests/github-action-helper.sh
Comment (satoru-takeuchi) [reply tests/github-action-helper.sh:280]: I'll add other tests in another PR as you suggested. 

https://github.com/rook/kubectl-rook-ceph/pull/229#issuecomment-2084428627
> Lets not block this PR for it but good to have tests to cover all cases, we can open a new issue to add more tests to ensure it works for all the cases

 I'll create an issue to track this test improvement.

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:35]: this looks to be same as `storageProvisionerBetaAnnotation` do we need both?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:79]: This check seems to be wrong, we are checking annotation but not checking if its rbd PVC. can you please check

**Review by Madhu-1 — COMMENTED** [review]
@satoru-takeuchi Thanks for the recent update, i have 2 more commits. everything else LGTM.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: Oops, sorry. I meant not "volume.beta.kubernetes.io/storage-provisioner" but "ceph-ssd.rbd.csi.ceph.com" for `cephCSIRBDStorageProviderAnnotation`.

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: Yes, it's completely wrong. Sorry for the inconvenience. This code passed integration tests contingency.

What I want to do is verify whether the target PVC is really a RBD PVC by checking the existence of one of the the following annotations.

- volume.beta.kubernetes.io/storage-provisioner: "ceph-ssd.rbd.csi.ceph.com"
- volume.kubernetes.io/storage-provisioner: "ceph-ssd.rbd.csi.ceph.com"

Static PVC [doesn't have these annotations](https://github.com/ceph/ceph-csi/blob/devel/docs/static-pvc.md).

Comment (Madhu-1) [reply cmd/commands/flatten_rbd_pvc.go]: instead of annotation can we check the presence of `ImageName` or something or check if it is CSI PV not in-tree PV?

File: cmd/commands/flatten_rbd_pvc.go
Comment (Madhu-1) [inline cmd/commands/flatten_rbd_pvc.go:80]: i think we dont need to worry about the static PV as most of the cases they are backed by RBD image not clones.

**Review by Madhu-1 — APPROVED** [review]
@satoru-takeuchi Thanks for taking this up!. LGTM

File: cmd/commands/flatten_rbd_pvc.go
Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: OK, I'll update this PR tomorrorw.

Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go]: @satoru-takeuchi I found that we've already checked the existence of `ImageName` as follows. 

```go
		imageName, ok := pv.Spec.CSI.VolumeAttributes["imageName"]
		if !ok {
			logging.Fatal(fmt.Errorf("PV %s doesn't contains `imageName` in VolumeAttributes", pvName))
		}
```

In other words, we got the name of RBD image name from RBD PVC's `spec.csi.volumeAttributes.imageName`. Since this field only exists in RBD PVC, it can be said that we checked whether the target PVC is really an RBD PVC here.

So re-removing checking attribute is enough and I updated this PR as so. Could you check the updated PR again? Correct me if I'm wrong.

File: cmd/commands/flatten_rbd_pvc.go
Comment (subhamkrai) [inline cmd/commands/flatten_rbd_pvc.go:59]: @satoru-takeuchi overall pr looks good, small suggestion if you could move the implementation function to under `pkg/` folder and not put it in the same file as where command is written. Thanks after that PR will be ready to merge.

Comment (satoru-takeuchi) [reply cmd/commands/flatten_rbd_pvc.go:39]: @subhamkrai Indeed, moved.

**Review by subhamkrai — APPROVED** [review]

File: cmd/commands/flatten_rbd_pvc.go
Comment (subhamkrai) [reply cmd/commands/flatten_rbd_pvc.go:39]: merged

**Pull request merged**
